// _project/entity-diff.js
'use strict'

const log = require('debug')('moysklad-couch-sync')
const jsondiffpatch = require('jsondiffpatch')

module.exports = jsondiffpatch.create({
  // used to match objects when diffing arrays, by default only === operator is used
  objectHash: function (entity) {
    // this function is used only to when objects are not equal by ref
    return entity.uuid
  },
  arrays: {
    // default true, detect items moved inside the array
    // (otherwise they will be registered as remove+add)
    detectMove: true,
    // default false, the value of items moved is not included in deltas
    includeValueOnMove: false
  },
  textDiff: {
    // default 60, minimum string length (left and right sides)
    // to use text diff algorythm: google-diff-match-patch
    minLength: 60
  },
  propertyFilter: function (name, context) {
    // this optional function can be specified to ignore object properties (eg. volatile data)
    // name: property name, present in either context.left or context.right objects
    // context: the diff context (has context.left and context.right objects)

    var obj
    var sides = ['left', 'right']

    // TODO Ad-hoc решение ..
    // Не правильно мутировать исходные объекты в этом месте, но так как подобное преобразование
    // лучше сделать внутри Jsonix, оставим так, пока не реализована эта возможность.
    if (name === 'contents') {
      return sides
        .map(side => {
          obj = context[side]
          if (obj.filename && obj.filename.slice(-5) === '.json') {
            if (typeof obj.contents === 'string') {
              try {
                obj.contents = JSON.parse(Buffer.from(obj.contents, 'base64').toString('utf8'))
              } catch (e) {
                log(`${obj.filename} file content parse failed:`, e.message)
              }
            }
            return true
          } else {
            return false
          }
        })
        .some(item => item)
    } else {
      // Изменения в updatedBy создают слишком большой diff (вынесего в родительский объект)
      return ['_id', '_rev', 'updatedBy'].indexOf(name) === -1
    }
  },
  cloneDiffValues: false /* default false. if true, values in the obtained delta will be cloned,
   to ensure delta keeps no references to left or right objects. this becomes useful
   if you're diffing and patching the same objects multiple times without serializing deltas.
   instead of true, a function can be specified here to provide a custom clone(value)
   */
})
